packages/authentication/docs/authentication-strategy.md

```diff
@@ -1,20 +1,16 @@
-### Auth strategy interface
+### Authentication strategy interface
 
 ```ts
 import {Request} from '@loopback/rest';
 
-interface AuthenticationStrategy {
-  // The resolver will read the `options` object from metadata, then invoke the
-  // `authenticate` with `options` if it exists.
-  authenticate(
-    request: Request,
-    options: object,
-  ): Promise<UserProfile | undefined>;
-
-  // This is a private function that extracts credential fields from a request,
-  // it is called in function `authenticate`. You could organize the extraction
-  // logic in this function or write them in `authenticate` directly without defining
-  // this extra utility.
-  private extractCredentials?(request: Request): Promise<Credentials>;
+export interface AuthenticationStrategy {
+  name: string;
+  authenticate(request: Request): Promise<UserProfile>;
```

Comment:
At the moment, `authenticate` can return either a user profile or `undefined`. I presume `undefined` is indicating there were no credentials provided by the client and thus the request is anonymous. (It's important to distinguish this case from the situation where the client provided some credentials but they were not valid - an error should be thrown in such case.)

In your new proposal, authentication strategies have to either return a user profile or throw an error.

How do you envision to handle anonymous requests with no credentials provided by the client?

I am not familiar with all details of the proposed authentication design. If anonymous requests are meant to be handled elsewhere in the new authentication solution, then please ignore my comment.

Reply:
@bajtos , what is meant by `anonymous` request?  In the controller classes, there are endpoints that won't be decorated with `@authenticate( '{authentication strategy name}'  )`, and some endpoints that will be decorated with it.  The authenticate action in the sequence will not attempt any authentication on an endpoint which doesn't have an authentication strategy specified. If a strategy is provided, that strategy will return a user if and only if the user credentials are valid, otherwise the strategy will throw an http error.

Reply:
By `anonymous` request I mean a request that does not include any credentials. For example, by opening http://localhost:3000/orders in your browser.

Let's say this that the controller method implementing `/orders` is decorated with `@authenticate(/*some strategy*/)`.

In the code we have so far, the authentication strategy returns `undefined`, meaning there are no credentials provided for the request, and it's up to `authenticate()` sequence action or even the sequence itself to decide how to handle the situation when an anonymous request is received for a resource that requires authentication.

IIUC your proposal correctly, you are pushing the responsibility of handling this situation to individual authentication strategies.

I have two concerns about that approach:
 - Different strategy can throw different HTTP error. Applications using multiple strategies can end up with inconsistent error responses depending on which strategy was applied.
 - It's more difficult to implement fallbacks, i.e. configure the app so that if one strategy did not find any credentials in the request, then we can try a different strategy to see if it's more successful.

Anyhow, I am not familiar with all details of the solution you are aiming for. If the change of `authenticate` signature is in line with your new overall design, then I am fine with that. As long as it's a conscious decision made after considering pros and cons of different approaches.

Reply:
@emonddr Can we keep the `undefined` in the return signature of the `authenticate` method given the above discussion and the points made by @bajtos? I am also leaning towards letting the sequence action or sequence handle anonymous requests in a unified way as opposed to delegating it to authentication strategies.

Reply:
Currently the shopping cart `sequence` looks like this:

```ts
export class MySequence implements SequenceHandler {
  constructor(
    @inject(SequenceActions.FIND_ROUTE) protected findRoute: FindRoute,
    @inject(SequenceActions.PARSE_PARAMS) protected parseParams: ParseParams,
    @inject(SequenceActions.INVOKE_METHOD) protected invoke: InvokeMethod,
    @inject(SequenceActions.SEND) public send: Send,
    @inject(SequenceActions.REJECT) public reject: Reject,
    @inject(AuthenticationBindings.AUTH_ACTION)
    protected authenticateRequest: AuthenticateFn,
  ) {}

  async handle(context: RequestContext) {
    try {

      const {request, response} = context;
      const route = this.findRoute(request);
      await this.authenticateRequest(request);
      const args = await this.parseParams(request, route);
      const result = await this.invoke(route, args);
      this.send(response, result);
    } catch (err) {
      this.reject(context, err);
    }
  }
}
```
Currently the shopping cart `authentication` action in the sequence looks like this:

```ts
export class AuthenticateActionProvider implements Provider<AuthenticateFn> {
  constructor(
    // The provider is instantiated for Sequence constructor,
    // at which time we don't have information about the current
    // route yet. This information is needed to determine
    // what auth strategy should be used.
    // To solve this, we are injecting a getter function that will
    // defer resolution of the strategy until authenticate() action
    // is executed.
    @inject.getter(AuthenticationBindings.STRATEGY)
    readonly getStrategy: Getter<AuthenticationStrategy>,
    @inject.setter(AuthenticationBindings.CURRENT_USER)
    readonly setCurrentUser: Setter<UserProfile>,
  ) {}

  /**
   * @returns authenticateFn
   */
  value(): AuthenticateFn {
    return request => this.action(request);
  }

  /**
   * The implementation of authenticate() sequence action.
   * @param request The incoming request provided by the REST layer
   */
  async action(request: Request): Promise<UserProfile | undefined> {
    const strategy = await this.getStrategy();
    if (!strategy) {
      // The invoked operation does not require authentication.
      return undefined;
    }
    if (!strategy.authenticate) {
      throw new Error('invalid strategy parameter');
    }
    const user = await strategy.authenticate(request);
    if (user) this.setCurrentUser(user);
    return user;
  }
}
```

Reply:
@jannyHou 's new `authentication` action for `authentication` package will look similar. See : https://github.com/strongloop/loopback-next/issues/2467 .


In the `authentication` action,
If a request comes through from an endpoint that is not decorated with a strategy(in controller via @authenticate('jwt') for example), then the action won't find a strategy and returns. 

If a request comes throw that does have a strategy associated with it, that strategy's authenticate method is called and returns a user profile which is placed on the context. (Or it throws an error, and that error percolates up the try/catch in the sequence which has an error handler. I don't see the point for an authentication strategy to return `undefined`.

Reply:
Are you saying that you want the `authenticate` method of a strategy to return { user: {value or undefined}, error: {value or undefined}}, and never throws an error, and it is up to the action to standardize the error handling in its block of code (regardless of strategy)?  In this case, still... the strategy.authenticate will always return an object, but it is the action that will handle throwing errors.

Reply:
I am asking to preserve the signature used by the current strategy interface in the shopping cart.

```ts
export interface AuthenticationStrategy {
  name: string;
  authenticate(request: Request): Promise<UserProfile | undefined>;
}
```

Take a look at the implementation of AuthenticateFn you posted above - it's already prepared to handle the case when the authentication strategy did not find any user credentials.

```ts
    const user = await strategy.authenticate(request);
    if (user) this.setCurrentUser(user);
    return user;
```